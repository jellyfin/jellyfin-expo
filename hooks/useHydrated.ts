/**
 * Copyright (c) 2025 Jellyfin Contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { useEffect, useState } from 'react';

import { useDownloadStore } from '../stores/DownloadStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

export const useIsHydrated = () => {
	const [ isHydrated, setIsHydrated ] = useState(false);
	const [ isRootStoreHydrated, setIsRootStoreHydrated ] = useState(false);
	const [ isDownloadStoreHydrated, setIsDownloadStoreHydrated ] = useState(false);
	const [ isServerStoreHydrated, setIsServerStoreHydrated ] = useState(false);
	const [ isSettingStoreHydrated, setIsSettingStoreHydrated ] = useState(false);

	useEffect(() => {
		// Root Store
		const unsubRootFinishHydration = useRootStore.persist.onFinishHydration(() => setIsRootStoreHydrated(true));
		setIsRootStoreHydrated(useRootStore.persist.hasHydrated());
		// Download Store
		const unsubDownloadFinishHydration = useDownloadStore.persist.onFinishHydration(() => setIsDownloadStoreHydrated(true));
		setIsDownloadStoreHydrated(useDownloadStore.persist.hasHydrated());
		// Server Store
		const unsubServerFinishHydration = useServerStore.persist.onFinishHydration((state) => {
			console.debug('>>> hydrated state', state);
			setIsServerStoreHydrated(true);
		});
		setIsServerStoreHydrated(useServerStore.persist.hasHydrated());
		// Settings Store
		const unsubSettingFinishHydration = useSettingStore.persist.onFinishHydration(() => setIsSettingStoreHydrated(true));
		setIsSettingStoreHydrated(useSettingStore.persist.hasHydrated());

		return () => {
			unsubRootFinishHydration();
			unsubDownloadFinishHydration();
			unsubServerFinishHydration();
			unsubSettingFinishHydration();
		};
	}, []);

	useEffect(() => {
		console.debug('Hydration state', isRootStoreHydrated, isDownloadStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated);
		if (isRootStoreHydrated && isDownloadStoreHydrated && isServerStoreHydrated && isSettingStoreHydrated) {
			setIsHydrated(true);
		}
	}, [ isDownloadStoreHydrated, isRootStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated ]);

	return isHydrated;
};
