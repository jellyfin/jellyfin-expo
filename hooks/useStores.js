/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { useDownloadStore } from '../stores/DownloadStore';
import { useMediaStore } from '../stores/MediaStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

// Compatibility for zustand conversion
export const useStores = () => ({
	rootStore: useRootStore(),
	downloadStore: useDownloadStore(),
	serverStore: useServerStore(),
	mediaStore: useMediaStore(),
	settingStore: useSettingStore()
});
