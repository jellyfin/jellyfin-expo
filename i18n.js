/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as Localization from 'expo-localization';
import i18next from 'i18next';
import { initReactI18next } from 'react-i18next';
import { I18nManager } from 'react-native';

import ar from './langs/ar.json';
import be from './langs/be.json';
import bg from './langs/bg.json';
import ca from './langs/ca.json';
import cs from './langs/cs.json';
import cy from './langs/cy.json';
import da from './langs/da.json';
import de from './langs/de.json';
import el from './langs/el.json';
import en from './langs/en.json';
import eo from './langs/eo.json';
import es from './langs/es.json';
import es_419 from './langs/es_419.json';
import es_AR from './langs/es_AR.json';
import et from './langs/et.json';
import fa from './langs/fa.json';
import fi from './langs/fi.json';
import fil from './langs/fil.json';
import fr from './langs/fr.json';
import ga from './langs/ga.json';
import he from './langs/he.json';
import hr from './langs/hr.json';
import hu from './langs/hu.json';
import hy from './langs/hy.json';
import id from './langs/id.json';
import is from './langs/is.json';
import it from './langs/it.json';
import ja from './langs/ja.json';
import kk from './langs/kk.json';
import ko from './langs/ko.json';
import lt from './langs/lt.json';
import lv from './langs/lv.json';
import mn from './langs/mn.json';
import ms from './langs/ms.json';
import mt from './langs/mt.json';
import nb_NO from './langs/nb_NO.json';
import nds from './langs/nds.json';
import nl from './langs/nl.json';
import pl from './langs/pl.json';
import pt from './langs/pt.json';
import pt_BR from './langs/pt_BR.json';
import pt_PT from './langs/pt_PT.json';
import ro from './langs/ro.json';
import ru from './langs/ru.json';
import sk from './langs/sk.json';
import sl from './langs/sl.json';
import sq from './langs/sq.json';
import sr from './langs/sr.json';
import sv from './langs/sv.json';
import ta from './langs/ta.json';
import tr from './langs/tr.json';
import uk from './langs/uk.json';
import vi from './langs/vi.json';
import zh_Hans from './langs/zh_Hans.json';
import zh_Hant from './langs/zh_Hant.json';

export const resources = {
	en: { translation: en },
	ar: { translation: ar },
	be: { translation: be },
	bg: { translation: bg },
	ca: { translation: ca },
	cs: { translation: cs },
	cy: { translation: cy },
	da: { translation: da },
	de: { translation: de },
	el: { translation: el },
	eo: { translation: eo },
	es: { translation: es },
	'es-419': { translation: es_419 },
	'es-AR': { translation: es_AR },
	et: { translation: et },
	fa: { translation: fa },
	fi: { translation: fi },
	fil: { translation: fil },
	fr: { translation: fr },
	ga: { translation: ga },
	he: { translation: he },
	hr: { translation: hr },
	hu: { translation: hu },
	hy: { translation: hy },
	id: { translation: id },
	is: { translation: is },
	it: { translation: it },
	ja: { translation: ja },
	kk: { translation: kk },
	ko: { translation: ko },
	lt: { translation: lt },
	lv: { translation: lv },
	mn: { translation: mn },
	ms: { translation: ms },
	mt: { translation: mt },
	'nb-NO': { translation: nb_NO },
	nds: { translation: nds },
	nl: { translation: nl },
	pl: { translation: pl },
	pt: { translation: pt },
	'pt-BR': { translation: pt_BR },
	'pt-PT': { translation: pt_PT },
	ro: { translation: ro },
	ru: { translation: ru },
	sk: { translation: sk },
	sl: { translation: sl },
	sq: { translation: sq },
	sr: { translation: sr },
	sv: { translation: sv },
	ta: { translation: ta },
	tr: { translation: tr },
	uk: { translation: uk },
	vi: { translation: vi },
	'zh-Hans': { translation: zh_Hans },
	'zh-Hant': { translation: zh_Hant }
};

// Ensure RTL layout is enabled for RTL locales
I18nManager.forceRTL(Localization.isRTL);

i18next
	.use(initReactI18next)
	.init({
		// debug: true,
		fallbackLng: 'en',
		// Use the fallback language for empty strings
		returnEmptyString: false,
		lng: Localization.locale,
		interpolation: {
			escapeValue: false
		},
		resources
	});

// Export i18next instance for use in tests
export default i18next;
