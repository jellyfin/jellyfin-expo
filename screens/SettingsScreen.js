/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useNavigation } from '@react-navigation/native';
import compareVersions from 'compare-versions';
import React, { useContext, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { Alert, Platform, SectionList, StyleSheet, View } from 'react-native';
import { Text, ThemeContext } from 'react-native-elements';
import { SafeAreaView } from 'react-native-safe-area-context';

import AppInfoFooter from '../components/AppInfoFooter';
import BrowserListItem from '../components/BrowserListItem';
import ButtonListItem from '../components/ButtonListItem';
import ServerListItem from '../components/ServerListItem';
import SwitchListItem from '../components/SwitchListItem';
import Links from '../constants/Links';
import Screens from '../constants/Screens';
import { useStores } from '../hooks/useStores';
import { isSystemThemeSupported } from '../utils/Device';

const SettingsScreen = () => {
	const { rootStore, serverStore, settingStore, mediaStore, downloadStore } = useStores();
	const navigation = useNavigation();
	const { t } = useTranslation();
	const { theme } = useContext(ThemeContext);

	useEffect(() => {
		// Fetch server info
		serverStore.fetchInfo();
	}, []);

	const onAddServer = () => {
		navigation.navigate(Screens.AddServerScreen);
	};

	const onDeleteServer = index => {
		Alert.alert(
			t('alerts.deleteServer.title'),
			t('alerts.deleteServer.description', { serverName: serverStore.servers[index]?.name }),
			[
				{ text: t('common.cancel') },
				{
					text: t('alerts.deleteServer.confirm'),
					onPress: () => {
						// Remove server and update active server
						serverStore.removeServer(index);
						settingStore.set({ activeServer: 0 });

						if (serverStore.servers.length > 0) {
							// More servers exist, navigate home
							navigation.replace(Screens.HomeScreen);
							navigation.navigate(Screens.HomeTab);
						} else {
							// No servers are present, navigate to add server screen
							navigation.replace(Screens.AddServerScreen);
						}
					},
					style: 'destructive'
				}
			]
		);
	};

	const onSelectServer = (index) => {
		settingStore.set({ activeServer: index });
		navigation.replace(Screens.HomeScreen);
		navigation.navigate(Screens.HomeTab);
	};

	const onResetApplication = () => {
		Alert.alert(
			t('alerts.resetApplication.title'),
			t('alerts.resetApplication.description'),
			[
				{ text: t('common.cancel') },
				{
					text: t('alerts.resetApplication.confirm'),
					onPress: () => {
						// Reset data in stores
						mediaStore.reset();
						downloadStore.reset();
						serverStore.reset();
						settingStore.reset();
						rootStore.reset();

						AsyncStorage.clear();

						// Navigate to the loading screen
						navigation.replace(Screens.AddServerScreen);
					},
					style: 'destructive'
				}
			]
		);
	};

	const AugmentedServerListItem = (props) => (
		<ServerListItem
			{...props}
			activeServer={settingStore.activeServer}
			onDelete={onDeleteServer}
			onPress={onSelectServer}
		/>
	);

	const getSections = () => {
		const settingsData = [{
			key: 'keep-awake-switch',
			title: t('settings.keepAwake'),
			value: settingStore.isScreenLockEnabled,
			onValueChange: (value) => settingStore.set({ isScreenLockEnabled: value })
		}];

		// Orientation lock is not supported on iPad without disabling multitasking
		// https://docs.expo.io/versions/latest/sdk/screen-orientation/#warning
		if (Platform.OS !== 'ios' || !Platform.isPad) {
			settingsData.push({
				key: 'rotation-lock-switch',
				title: t('settings.rotationLock'),
				value: settingStore.isRotationLockEnabled,
				onValueChange: (value) => settingStore.set({ isRotationLockEnabled: value })
			});
		}

		const playbackSettingsData = [];

		// TODO: Add Android support for native video player
		if (Platform.OS === 'ios') {
			playbackSettingsData.push({
				key: 'native-video-switch',
				title: t('settings.nativeVideoPlayer'),
				badge: {
					value: t('common.beta')
				},
				value: settingStore.isNativeVideoPlayerEnabled,
				onValueChange: (value) => {
					settingStore.set({ isNativeVideoPlayerEnabled: value });
					rootStore.set({ isReloadRequired: true });
				}
			});

			if (compareVersions.compare(Platform.Version, '12', '>')) {
				playbackSettingsData.push({
					key: 'native-video-fmp4-switch',
					title: t('settings.fmp4Support'),
					value: settingStore.isFmp4Enabled,
					disabled: !settingStore.isNativeVideoPlayerEnabled,
					onValueChange: (value) => {
						settingStore.set({ isFmp4Enabled: value });
						rootStore.set({ isReloadRequired: true });
					}
				});
			}
		}

		const appearanceSettingsData = [{
			key: 'tab-labels-switch',
			title: t('settings.tabLabels'),
			value: settingStore.isTabLabelsEnabled,
			onValueChange: (value) => settingStore.set({ isTabLabelsEnabled: value })
		}];

		if (isSystemThemeSupported()) {
			appearanceSettingsData.push({
				key: 'system-theme-switch',
				title: t('settings.systemTheme'),
				value: settingStore.isSystemThemeEnabled,
				onValueChange: (value) => settingStore.set({ isSystemThemeEnabled: value })
			});
		}

		// TODO: This should be able to select from a list not just a switch
		appearanceSettingsData.push({
			key: 'theme-switch',
			title: t('settings.lightTheme'),
			disabled: settingStore.isSystemThemeEnabled,
			value: settingStore.themeId === 'light',
			onValueChange: (value) => settingStore.set({ themeId: value ? 'light' : 'dark' })
		});

		return [
			{
				title: t('headings.servers'),
				data: serverStore.servers.slice(),
				keyExtractor: (item, index) => `server-${index}`,
				renderItem: AugmentedServerListItem
			},
			{
				title: t('headings.addServer'),
				hideHeader: true,
				data: [{
					key: 'add-server-button',
					title: t('headings.addServer'),
					onPress: onAddServer
				}],
				renderItem: ButtonListItem
			},
			{
				title: t('headings.settings'),
				data: settingsData,
				renderItem: SwitchListItem
			},
			{
				title: t('headings.playback'),
				data: playbackSettingsData,
				renderItem: SwitchListItem
			},
			{
				title: t('headings.appearance'),
				data: appearanceSettingsData,
				renderItem: SwitchListItem
			},
			{
				title: t('headings.links'),
				data: Links.map(link => ({
					...link,
					name: t(link.name)
				})),
				renderItem: BrowserListItem
			},
			{
				title: t('alerts.resetApplication.title'),
				hideHeader: true,
				data: [{
					key: 'reset-app-button',
					title: t('alerts.resetApplication.title'),
					titleStyle: {
						color: theme.colors.error
					},
					onPress: onResetApplication
				}],
				renderItem: ButtonListItem
			}
		];
	};

	return (
		<SafeAreaView
			style={{
				...styles.container,
				backgroundColor: theme.colors.background
			}}
			edges={[ 'right', 'left' ]}
		>
			<SectionList
				sections={getSections()}
				extraData={{
					activeServer: settingStore.activeServer,
					isFetching: serverStore.fetchInfo.pending
				}}
				renderItem={({ item }) => <Text>{JSON.stringify(item)}</Text>}
				renderSectionHeader={({ section: { data, title, hideHeader } }) => {
					if (!data || data.length === 0) {
						return null;
					}
					if (hideHeader) {
						return <View style={styles.emptyHeader} />;
					}
					return (
						<Text style={{
							...styles.header,
							backgroundColor: theme.colors.background,
							color: theme.colors.grey1
						}}>{title}</Text>
					);
				}}
				renderSectionFooter={({ section: { data } }) => {
					if (!data || data.length === 0) {
						return null;
					}
					return <View style={styles.footer} />;
				}}
				ListFooterComponent={AppInfoFooter}
				showsVerticalScrollIndicator={false}
			/>
		</SafeAreaView>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1
	},
	header: {
		fontWeight: '600',
		paddingVertical: 8,
		paddingHorizontal: 15,
		marginBottom: 1
	},
	emptyHeader: {
		marginTop: 15
	},
	footer: {
		marginBottom: 15
	}
});

export default SettingsScreen;
