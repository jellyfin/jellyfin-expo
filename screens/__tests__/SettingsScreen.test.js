/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { NavigationContainer } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React from 'react';
import { Platform } from 'react-native';
import { ThemeProvider } from 'react-native-elements';

import '../../i18n';
import { useStores } from '../../hooks/useStores';
import SettingsScreen from '../SettingsScreen';

// Mock Platform to allow testing platform specific code
jest.mock('react-native/Libraries/Utilities/Platform');
// Button fails to render in some cases so it needs mocked
jest.mock('react-native-elements/dist/buttons/Button', () => 'Button');

jest.mock('../../hooks/useStores');
useStores.mockImplementation(() => ({
	rootStore: {},
	settingStore: {},
	serverStore: {
		fetchInfo: jest.fn(),
		servers: []
	}
}));

describe('SettingsScreen', () => {
	it('should render correctly', () => {
		Platform.Version = '13';

		const { toJSON } = render(
			<ThemeProvider>
				<NavigationContainer>
					<SettingsScreen />
				</NavigationContainer>
			</ThemeProvider>
		);

		expect(toJSON()).toMatchSnapshot();
	});
});
