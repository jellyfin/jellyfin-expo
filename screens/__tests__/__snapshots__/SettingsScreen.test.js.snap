// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "off",
      "left": "additive",
      "right": "additive",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    data={
      [
        {
          "data": [],
          "keyExtractor": [Function],
          "renderItem": [Function],
          "title": "Servers",
        },
        {
          "data": [
            {
              "key": "add-server-button",
              "onPress": [Function],
              "title": "Add Server",
            },
          ],
          "hideHeader": true,
          "renderItem": [Function],
          "title": "Add Server",
        },
        {
          "data": [
            {
              "key": "keep-awake-switch",
              "onValueChange": [Function],
              "title": "Keep Screen Awake",
              "value": undefined,
            },
            {
              "key": "rotation-lock-switch",
              "onValueChange": [Function],
              "title": "Rotation Lock",
              "value": undefined,
            },
          ],
          "renderItem": [Function],
          "title": "Settings",
        },
        {
          "data": [
            {
              "badge": {
                "value": "BETA",
              },
              "key": "native-video-switch",
              "onValueChange": [Function],
              "title": "Use Native Video Player",
              "value": undefined,
            },
            {
              "disabled": true,
              "key": "native-video-fmp4-switch",
              "onValueChange": [Function],
              "title": "Prefer fMP4 in HLS",
              "value": undefined,
            },
          ],
          "renderItem": [Function],
          "title": "Playback",
        },
        {
          "data": [
            {
              "key": "tab-labels-switch",
              "onValueChange": [Function],
              "title": "Show Tab Labels",
              "value": undefined,
            },
            {
              "key": "system-theme-switch",
              "onValueChange": [Function],
              "title": "Use System Theme",
              "value": undefined,
            },
            {
              "disabled": undefined,
              "key": "theme-switch",
              "onValueChange": [Function],
              "title": "Use Light Theme",
              "value": false,
            },
          ],
          "renderItem": [Function],
          "title": "Appearance",
        },
        {
          "data": [
            {
              "icon": {
                "name": "globe-outline",
                "type": "ionicon",
              },
              "key": "links-website",
              "name": "Jellyfin Website",
              "url": "https://jellyfin.org/",
            },
            {
              "icon": {
                "name": "book",
                "type": "ionicon",
              },
              "key": "links-documentation",
              "name": "Documentation",
              "url": "https://jellyfin.org/docs/",
            },
            {
              "icon": {
                "name": "logo-github",
                "type": "ionicon",
              },
              "key": "links-source",
              "name": "Source Code",
              "url": "https://github.com/jellyfin/jellyfin-expo",
            },
            {
              "icon": {
                "name": "translate",
                "type": "material",
              },
              "key": "links-translate",
              "name": "Translate",
              "url": "https://translate.jellyfin.org/projects/jellyfin/jellyfin-expo/",
            },
            {
              "icon": {
                "name": "help-buoy",
                "type": "ionicon",
              },
              "key": "links-help",
              "name": "Getting Help",
              "url": "https://jellyfin.org/docs/general/getting-help",
            },
          ],
          "renderItem": [Function],
          "title": "Links",
        },
        {
          "data": [
            {
              "key": "reset-app-button",
              "onPress": [Function],
              "title": "Reset Application",
              "titleStyle": {
                "color": "#ff190c",
              },
            },
          ],
          "hideHeader": true,
          "renderItem": [Function],
          "title": "Reset Application",
        },
      ]
    }
    extraData={
      {
        "activeServer": undefined,
        "isFetching": undefined,
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={
      [
        0,
        2,
        5,
        9,
      ]
    }
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginTop": 15,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          replaceTheme={[Function]}
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "borderTopWidth": 0.5,
                "flexDirection": "row",
              }
            }
            testID="padView"
          >
            <View
              replaceTheme={[Function]}
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
              updateTheme={[Function]}
            >
              <Button
                buttonStyle={
                  {
                    "padding": 0,
                  }
                }
                onPress={[Function]}
                testID="button"
                title="Add Server"
                titleStyle={
                  {
                    "textAlign": "auto",
                    "width": "100%",
                  }
                }
                type="clear"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 15,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            {
              "backgroundColor": undefined,
              "color": "#43484d",
              "fontWeight": "600",
              "marginBottom": 1,
              "paddingHorizontal": 15,
              "paddingVertical": 8,
            }
          }
        >
          Settings
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          replaceTheme={[Function]}
          testID="switch-list-item"
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "borderTopWidth": 0.5,
                "flexDirection": "row",
              }
            }
            testID="padView"
          >
            <View
              replaceTheme={[Function]}
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
              updateTheme={[Function]}
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    {
                      "backgroundColor": "transparent",
                      "color": "#242424",
                    }
                  }
                  testID="title"
                >
                  Keep Screen Awake
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <RCTSwitch
              accessibilityRole="switch"
              onChange={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 31,
                  "width": 51,
                }
              }
              testID="switch"
              value={false}
            />
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          replaceTheme={[Function]}
          testID="switch-list-item"
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
              }
            }
            testID="padView"
          >
            <View
              replaceTheme={[Function]}
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
              updateTheme={[Function]}
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    {
                      "backgroundColor": "transparent",
                      "color": "#242424",
                    }
                  }
                  testID="title"
                >
                  Rotation Lock
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <RCTSwitch
              accessibilityRole="switch"
              onChange={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 31,
                  "width": 51,
                }
              }
              testID="switch"
              value={false}
            />
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 15,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            {
              "backgroundColor": undefined,
              "color": "#43484d",
              "fontWeight": "600",
              "marginBottom": 1,
              "paddingHorizontal": 15,
              "paddingVertical": 8,
            }
          }
        >
          Playback
        </Text>
      </View>
      <View
        style={
          {
            "height": 0,
          }
        }
      />
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "margin": 15,
            }
          }
        >
          <Text
            onLongPress={[Function]}
            style={
              {
                "color": "#43484d",
                "fontSize": 15,
              }
            }
            testID="app-name"
          >
            Jellyfin Mobile (mock)
          </Text>
          <Text
            style={
              {
                "color": "#43484d",
                "fontSize": 15,
              }
            }
            testID="app-version"
          >
            undefined (undefined)
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
